app/code/Magento/CatalogGraphQl/Model/Resolver/Product/SmallImageUrl.php

```diff
@@ -0,0 +1,131 @@
+use Magento\Framework\GraphQl\Query\Resolver\Value;
+use Magento\Framework\GraphQl\Query\Resolver\ValueFactory;
+use Magento\Framework\GraphQl\Query\ResolverInterface;
+use Magento\Framework\GraphQl\Schema\Type\ResolveInfo;
+use Magento\Store\Model\StoreManagerInterface;
+
+/**
+ * Returns product's small image. If the small image is not set, returns a placeholder
+ */
+class SmallImageUrl implements ResolverInterface
+{
+    /**
+     * @var GalleryResourceFactory
+     */
+    private $galleryResourceFactory;
+
+    /**
+     * @var AreaList
+     */
+    private $areaList;
+
+    /**
+     * @var CatalogImageHelper
+     */
+    private $catalogImageHelper;
+
+    /**
+     * @var ValueFactory
+     */
+    private $valueFactory;
+
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * @param ValueFactory $valueFactory
+     * @param CatalogImageHelper $catalogImageHelper
+     * @param AreaList $areaList
+     * @param GalleryResourceFactory $galleryResourceFactory
+     * @param StoreManagerInterface $storeManager
+     */
+    public function __construct(
+        ValueFactory $valueFactory,
+        CatalogImageHelper $catalogImageHelper,
+        AreaList $areaList,
+        GalleryResourceFactory $galleryResourceFactory,
+        StoreManagerInterface $storeManager
+    ) {
+        $this->valueFactory = $valueFactory;
+        $this->catalogImageHelper = $catalogImageHelper;
+        $this->areaList = $areaList;
+        $this->galleryResourceFactory = $galleryResourceFactory;
+        $this->storeManager = $storeManager;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function resolve(
+        Field $field,
+        $context,
+        ResolveInfo $info,
+        array $value = null,
+        array $args = null
+    ): Value {
+        if (!isset($value['model'])) {
+            $result = function () {
+                return null;
+            };
+            return $this->valueFactory->create($result);
+        }
+        /** @var Product $product */
+        $product = $value['model'];
+
+        /* If small_image is not loaded for product, need to load it separately */
+        if (!$product->getSmallImage()) {
+            $galleryResource = $this->galleryResourceFactory->create();
+            $currentStoreId =$this->storeManager->getStore()->getId();
+            $productImages = $galleryResource->getProductImages($product, [$currentStoreId]);
+            $productSmallImage = $this->getSmallImageFromGallery($productImages);
+            $product->setSmallImage($productSmallImage);
+        }
+
+        /* Design area is necessary to return the correct storefront image URL (or a placeholder) */
+        $area = $this->areaList->getArea(Area::AREA_FRONTEND);
+        $area->load(Area::PART_DESIGN);
+
+        $smallImageURL = $this->catalogImageHelper->init($product, 'product_small_image')->getUrl();
+        $product->getMediaAttributes();
```

Comment:
Why is this needed?

Reply:
We need this part to generate a correct URL to the image. The product media gallery does not contain any information about the product image URL (only relative path). Upon a product image URL generation, the system takes into account some factors (like static content signing). The loaded area is also taken into account (in the core image retrieving flow). That's why we need to use these operations.
By removing the area loading, for example, we will always have a path (incorrect) to the placeholder instead of image URL.

Reply:
Why we need `$product->getMediaAttributes()` call specifically? It looks like temporal coupling.

Reply:
Yes, you are right. It was an initial solution and I tried to reuse as much Magento core approaches as possible. Unfortunately, we have the temporal coupling here there in the core. I'm going to dig deeper and find more clear way for that. But, probably, we will need to introduce some new logic within a scope of this task for images URL generation that is different that we have in the code to avoid temporal coupling and other issues (especially performance issues).